app/service/sorting.go

```diff
@@ -0,0 +1,191 @@
+	sortStatements := strings.Split(sortingRules, ",")
+	usedFields = make([]string, 0, len(sortStatements)+1)
+	fieldsDirections = make(map[string]int, len(sortStatements)+1)
+	for _, sortStatement := range sortStatements {
+		fieldName, direction := getFieldNameAndDirectionFromSortStatement(sortStatement)
+		if fieldsDirections[fieldName] != 0 {
+			return nil, nil, fmt.Errorf("a field cannot be a sorting parameter more than once: %q", fieldName)
+		}
+		if _, ok := acceptedFields[fieldName]; !ok {
+			return nil, nil, fmt.Errorf("unknown field in sorting parameters: %q", fieldName)
+		}
+		fieldsDirections[fieldName] = direction
+		usedFields = append(usedFields, fieldName)
+	}
+	if fieldsDirections["id"] == 0 {
+		fieldsDirections["id"] = 1
+		usedFields = append(usedFields, "id")
+	}
+	return
+}
+
+func getFieldNameAndDirectionFromSortStatement(sortStatement string) (string, int) {
+	var direction int
+	if len(sortStatement) > 0 && sortStatement[0] == '-' {
+		sortStatement = sortStatement[1:]
+		direction = -1
+	} else {
+		direction = 1
+	}
+	fieldName := sortStatement
+	return fieldName, direction
+}
+
+func applyOrder(query *database.DB, usedFields []string, acceptedFields map[string]*FieldSortingParams,
+	fieldsDirections map[string]int) *database.DB {
+	usedFieldsNumber := len(usedFields)
+	orderStrings := make([]string, 0, usedFieldsNumber)
+	for _, field := range usedFields {
+		var directionString string
+		if fieldsDirections[field] == 1 {
+			directionString = "ASC"
+		} else {
+			directionString = "DESC"
+		}
+		orderStrings = append(orderStrings, acceptedFields[field].ColumnName+" "+directionString)
+	}
+	if len(orderStrings) > 0 {
+		query = query.Order(strings.Join(orderStrings, ", "))
+	}
+	return query
+}
+
+func parsePagingParameters(r *http.Request, usedFields []string,
+	acceptedFields map[string]*FieldSortingParams) ([]interface{}, error) {
+	fromValueSkipped := false
+	fromValueAccepted := false
+	fromValues := make([]interface{}, 0, len(usedFields))
+	for _, fieldName := range usedFields {
+		var value interface{}
+		fromFieldName := "from." + fieldName
+		if len(r.URL.Query()[fromFieldName]) > 0 {
+			var err error
+			value, err = getFromValueForField(r, fieldName, acceptedFields)
+			if err != nil {
+				return nil, err
+			}
+			fromValueAccepted = true
+		} else {
+			fromValueSkipped = true
+			continue
+		}
+		fromValues = append(fromValues, value)
+	}
+	if fromValueAccepted && fromValueSkipped {
+		fromParameters := strings.Join(usedFields, ", from.")
+		return nil, fmt.Errorf("all 'from' parameters (from.%s) or none of them must be present", fromParameters)
+	}
+	return fromValues, nil
+}
+
+func getFromValueForField(r *http.Request, fieldName string,
+	acceptedFields map[string]*FieldSortingParams) (interface{}, error) {
+	fromFieldName := "from." + fieldName
+	switch acceptedFields[fieldName].FieldType {
+	case "string":
+		return r.URL.Query()[fromFieldName][0], nil
```

Comment:
I think you are supposed to use `Get(...)` on `Values` (the result type from `Query()`). In particular, they check that the slice is not empty (I supposed it may happen).

(apply probably to other part of the code)

Reply:
We check that in parsePagingParameters(): ```
		if len(r.URL.Query()[fromFieldName]) > 0 {
```

Reply:
but they created this "Values" abstraction so that you do not do it in code by yourself, why not using it?

Reply:
This "Values" abstraction doesn't allow us to distinguish between absent values and empty values. In both cases it returns an empty string.